Mathlib/Analysis/InnerProductSpace/TwoDim.lean

```diff
@@ -0,0 +1,676 @@
+      areaForm_apply_self, neg_zero, smul_eq_mul, mul_zero, areaForm_rightAngleRotation_right,
+      real_inner_self_eq_norm_sq, zero_add, Real.rpow_two, mul_neg]
+    rw [o.areaForm_swap]
+    ring
+#align orientation.inner_mul_inner_add_area_form_mul_area_form' Orientation.inner_mul_inner_add_areaForm_mul_areaForm'
+
+/-- For vectors `a x y : E`, the identity `⟪a, x⟫ * ⟪a, y⟫ + ω a x * ω a y = ‖a‖ ^ 2 * ⟪x, y⟫`. -/
+theorem inner_mul_inner_add_areaForm_mul_areaForm (a x y : E) :
+    ⟪a, x⟫ * ⟪a, y⟫ + ω a x * ω a y = ‖a‖ ^ 2 * ⟪x, y⟫ :=
+  congr_arg (fun f : E →ₗ[ℝ] ℝ => f y) (o.inner_mul_inner_add_areaForm_mul_areaForm' a x)
+#align orientation.inner_mul_inner_add_area_form_mul_area_form Orientation.inner_mul_inner_add_areaForm_mul_areaForm
+
+theorem inner_sq_add_areaForm_sq (a b : E) : ⟪a, b⟫ ^ 2 + ω a b ^ 2 = ‖a‖ ^ 2 * ‖b‖ ^ 2 := by
+  simpa [sq, real_inner_self_eq_norm_sq] using o.inner_mul_inner_add_areaForm_mul_areaForm a b b
+#align orientation.inner_sq_add_area_form_sq Orientation.inner_sq_add_areaForm_sq
+
+/-- For vectors `a x y : E`, the identity `⟪a, x⟫ * ω a y - ω a x * ⟪a, y⟫ = ‖a‖ ^ 2 * ω x y`. (See
+`Orientation.inner_mul_areaForm_sub` for the "applied" form.) -/
+theorem inner_mul_areaForm_sub' (a x : E) : ⟪a, x⟫ • ω a - ω a x • innerₛₗ ℝ a = ‖a‖ ^ 2 • ω x := by
+  by_cases ha : a = 0
+  · simp [ha]
+  apply (o.basisRightAngleRotation a ha).ext
+  intro i
+  fin_cases i
+  · simp only [o.areaForm_swap a x, neg_smul, sub_neg_eq_add, Fin.mk_zero,
+      coe_basisRightAngleRotation, Matrix.cons_val_zero, LinearMap.add_apply, LinearMap.smul_apply,
+      areaForm_apply_self, smul_eq_mul, mul_zero, innerₛₗ_apply, real_inner_self_eq_norm_sq,
+      zero_add, Real.rpow_two]
+    ring
+  · simp only [Fin.mk_one, coe_basisRightAngleRotation, Matrix.cons_val_one, Matrix.head_cons,
+      LinearMap.sub_apply, LinearMap.smul_apply, areaForm_rightAngleRotation_right,
+      real_inner_self_eq_norm_sq, smul_eq_mul, innerₛₗ_apply, inner_rightAngleRotation_right,
+      areaForm_apply_self, neg_zero, mul_zero, sub_zero, Real.rpow_two, real_inner_comm]
+    ring
+#align orientation.inner_mul_area_form_sub' Orientation.inner_mul_areaForm_sub'
+
+/-- For vectors `a x y : E`, the identity `⟪a, x⟫ * ω a y - ω a x * ⟪a, y⟫ = ‖a‖ ^ 2 * ω x y`. -/
+theorem inner_mul_areaForm_sub (a x y : E) : ⟪a, x⟫ * ω a y - ω a x * ⟪a, y⟫ = ‖a‖ ^ 2 * ω x y :=
+  congr_arg (fun f : E →ₗ[ℝ] ℝ => f y) (o.inner_mul_areaForm_sub' a x)
+#align orientation.inner_mul_area_form_sub Orientation.inner_mul_areaForm_sub
+
+theorem nonneg_inner_and_areaForm_eq_zero_iff_sameRay (x y : E) :
+    0 ≤ ⟪x, y⟫ ∧ ω x y = 0 ↔ SameRay ℝ x y := by
+  by_cases hx : x = 0
+  · simp [hx]
+  constructor
+  · let a : ℝ := (o.basisRightAngleRotation x hx).repr y 0
+    let b : ℝ := (o.basisRightAngleRotation x hx).repr y 1
+    suffices 0 ≤ a * ‖x‖ ^ 2 ∧ b * ‖x‖ ^ 2 = 0 → SameRay ℝ x (a • x + b • J x) by
+      -- Porting note: `simp only` rewritten as `rw`
```

Comment:
Could you say why? We're eventually going to have to go back over all of these, so breadcrumbs are helpful!

Reply:
The suffices block was originally
```lean
    suffices 0 ≤ a * ‖x‖ ^ 2 ∧ b * ‖x‖ ^ 2 = 0 → SameRay ℝ x (a • x + b • J x) by
      rw [← (o.basisRightAngleRotation x hx).sum_repr y]
      simp only [Fin.sum_univ_succ, coe_basisRightAngleRotation, Matrix.cons_val_zero,
        Fin.succ_zero_eq_one', Fintype.univ_of_isEmpty, Finset.sum_empty, o.areaForm_apply_self,
        map_smul, map_add, map_zero, inner_smul_left, inner_smul_right, inner_add_left,
        inner_add_right, inner_zero_right, LinearMap.add_apply, Matrix.cons_val_one,
        Matrix.head_cons, Algebra.id.smul_eq_mul, o.areaForm_rightAngleRotation_right,
        MulZeroClass.mul_zero, add_zero, zero_add, neg_zero, o.inner_rightAngleRotation_right,
        o.areaForm_apply_self, real_inner_self_eq_norm_sq]
      exact this
```
After putting the 2220 hack on top of the file (and inserting a coercion – the 0 in `0 ≤ a` should be a real) the original proof works after all. In fact the simp line can be shortened:
```
    suffices ↑0 ≤ a * ‖x‖ ^ 2 ∧ b * ‖x‖ ^ 2 = 0 → SameRay ℝ x (a • x + b • J x) by
      rw [← (o.basisRightAngleRotation x hx).sum_repr y]
      simp only [Fin.sum_univ_succ, coe_basisRightAngleRotation, Matrix.cons_val_zero,
        Fin.succ_zero_eq_one', Fintype.univ_of_isEmpty, Finset.sum_empty, areaForm_apply_self,
        map_smul, map_add, real_inner_smul_right, inner_add_right, Matrix.cons_val_one,
        Matrix.head_cons, Algebra.id.smul_eq_mul, areaForm_rightAngleRotation_right,
        mul_zero, add_zero, zero_add, neg_zero, inner_rightAngleRotation_right,
        real_inner_self_eq_norm_sq]
      exact this
```